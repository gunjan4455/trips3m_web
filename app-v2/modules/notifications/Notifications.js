import React from 'react';
import PropTypes from 'prop-types';

import { GNotificationBell } from './G';
import config from 'appConfig';
import { getRtmsConnection, MessageType, RtmsChannels } from 'helpers/rtms';
import Modal from 'modules/shared/GenericModal';
import ModalHeaderCustom from 'modules/shared/headers/ModalHeaderCustom';
import NotificationBell from './components/NotificationBell';
import NotificationList from './components/NotificationList';

class Notifications extends React.Component {
  static propTypes = {
    userId: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string])
      .isRequired,
    unreadNotificationCount: PropTypes.number,
    messages: PropTypes.array,
    remainingMessages: PropTypes.number.isRequired,

    loadMoreNotifications: PropTypes.func.isRequired,
    addMessage: PropTypes.func.isRequired,
    fetchNotifications: PropTypes.func.isRequired,
    markNotificationRead: PropTypes.func.isRequired,
  };

  static defaultProps = {
    messages: [],
    unreadNotificationCount: undefined
  };

  constructor(props) {
    super(props);
    this.state = {};
    this.handleLoadMore = this.handleLoadMore.bind(this);
  }

  componentDidMount() {
    try {
      // load notifications
      this.props.fetchNotifications(this.props.userId);
      this.subscribeToNotification();
    } catch (e) {
      console.log(e);
    }
  }

  componentWillUnmount() {
    try {
      if (config.useRtms) {
        this.disposeExistingRtms();
      } else {
        this.unSubscribeNotification();
      }
    } catch (e) {
      console.log(e);
    }
  }

  unsubscribeRtms() {
    if (this.rtms_) {
      this.rtms_.close();
    }
  }

  async subscribeToRtms() {
    const { userId } = this.props;
    const channel = RtmsChannels.SYSTEM;
    this.rtms_ = await getRtmsConnection({ userId, channel });

    // Notification message has come
    this.rtms_.on(MessageType.NOTIFICATION, message => {
      const { payload: [{ content }] } = JSON.parse(message);
      this.props.addMessage(content);
    });
  }

  unSubscribeNotification() {
    this.unsubscribeRtms();
  }

  subscribeToNotification() {
    this.subscribeToRtms();
  }

  disposeExistingRtms = () => {
    // close web socket connection and remove all listners
    if (this.rtms_) {
      this.rtms_.close();
    }
  };

  handleLoadMore() {
    this.props.loadMoreNotifications();
  }

  trackSegmentEvent = (event) => {
    /*trackSegment({
      event: event,
      category: 'The concerned page',
      section: 'Global Menu',
      object: '',
      cta: 'Notification bell',
    });*/
  };

  render() {
    const notificationsPresent = (this.props.remainingMessages > 0);
    const {
      unreadNotificationCount, markNotificationRead, loadMoreNotifications,
      userId, messages, remainingMessages
    } = this.props;

    return (
      <GNotificationBell className="flex alignCenter justifyCenter at_nofification">
        <div className="notification-hover-box relative">
          <Modal
            trigger={<NotificationBell markNotificationRead={markNotificationRead}
                                       unreadNotificationCount={unreadNotificationCount}
                                       userId={userId} />}
            fullView={true}
            hasCustomClose={true}
          >
            <div className="notifications-head absolute wfull hfull flex flexDColumn pb48">
              <ModalHeaderCustom heading='Notifications' />
              <NotificationList
                userId={userId}
                messages={messages}
                notificationsPresent={notificationsPresent}
                loadMoreNotifications={loadMoreNotifications}
                remainingMessages={remainingMessages}
              />
            </div>
          </Modal>
        </div>
      </GNotificationBell>
    );
  }
}

export default Notifications;
