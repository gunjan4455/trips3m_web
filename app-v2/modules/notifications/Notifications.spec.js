import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

import Notifications from './Notifications';

configure({ adapter: new Adapter() });

const DEFAULT_PROPS = {
  userId: '',
  unreadNotificationCount: 0,
  messages: [],
  remainingMessages: 0,

  loadMoreNotifications: jest.fn,
  addMessage: jest.fn,
  fetchNotifications: jest.fn,
  markNotificationRead: jest.fn,
};

const shallowWrappperNotifications = (props) => {
  return shallow(<Notifications {...props} />);
};

describe('Notifications: Render Correctly ', () => {

  let props = { ...DEFAULT_PROPS };

  afterEach(() => {
    props = { ...DEFAULT_PROPS };
    jest.clearAllMocks();
  });

  it('should match with the previous snapshot', () => {
    const wrapper = shallowWrappperNotifications(props);
    expect(wrapper).toMatchSnapshot();
  });

});
