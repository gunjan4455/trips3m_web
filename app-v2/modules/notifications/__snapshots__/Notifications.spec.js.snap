// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications: Render Correctly  should match with the previous snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Notifications
    addMessage={[Function]}
    fetchNotifications={[Function]}
    loadMoreNotifications={[Function]}
    markNotificationRead={[Function]}
    messages={Array []}
    remainingMessages={0}
    unreadNotificationCount={0}
    userId=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="notification-hover-box relative"
      >
        <ModalComponent
          customClass=""
          customCloseTrigger={false}
          disabled={false}
          fullView={true}
          hasCustomClose={true}
          hasNoClose={false}
          isTriggerFullView={false}
          isWhiteIcon={false}
          onAfterModalOpen={[Function]}
          onRequestCloseCustom={[Function]}
          trigger={
            <NotificationBell
              markNotificationRead={[Function]}
              unreadNotificationCount={0}
              userId=""
            />
          }
        >
          <div
            className="notifications-head absolute wfull hfull flex flexDColumn pb48"
          >
            <ModalHeaderCustom
              heading="Notifications"
            />
            <NotificationList
              loadMoreNotifications={[Function]}
              messages={Array []}
              notificationsPresent={false}
              remainingMessages={0}
              userId=""
            />
          </div>
        </ModalComponent>
      </div>,
      "className": "flex alignCenter justifyCenter at_nofification",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ModalComponent
          customClass=""
          customCloseTrigger={false}
          disabled={false}
          fullView={true}
          hasCustomClose={true}
          hasNoClose={false}
          isTriggerFullView={false}
          isWhiteIcon={false}
          onAfterModalOpen={[Function]}
          onRequestCloseCustom={[Function]}
          trigger={
            <NotificationBell
              markNotificationRead={[Function]}
              unreadNotificationCount={0}
              userId=""
            />
          }
        >
          <div
            className="notifications-head absolute wfull hfull flex flexDColumn pb48"
          >
            <ModalHeaderCustom
              heading="Notifications"
            />
            <NotificationList
              loadMoreNotifications={[Function]}
              messages={Array []}
              notificationsPresent={false}
              remainingMessages={0}
              userId=""
            />
          </div>
        </ModalComponent>,
        "className": "notification-hover-box relative",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            className="notifications-head absolute wfull hfull flex flexDColumn pb48"
          >
            <ModalHeaderCustom
              heading="Notifications"
            />
            <NotificationList
              loadMoreNotifications={[Function]}
              messages={Array []}
              notificationsPresent={false}
              remainingMessages={0}
              userId=""
            />
          </div>,
          "customClass": "",
          "customCloseTrigger": false,
          "disabled": false,
          "fullView": true,
          "hasCustomClose": true,
          "hasNoClose": false,
          "isTriggerFullView": false,
          "isWhiteIcon": false,
          "onAfterModalOpen": [Function],
          "onRequestCloseCustom": [Function],
          "trigger": <NotificationBell
            markNotificationRead={[Function]}
            unreadNotificationCount={0}
            userId=""
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ModalHeaderCustom
                heading="Notifications"
              />,
              <NotificationList
                loadMoreNotifications={[Function]}
                messages={Array []}
                notificationsPresent={false}
                remainingMessages={0}
                userId=""
              />,
            ],
            "className": "notifications-head absolute wfull hfull flex flexDColumn pb48",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "heading": "Notifications",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "loadMoreNotifications": [Function],
                "messages": Array [],
                "notificationsPresent": false,
                "remainingMessages": 0,
                "userId": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
      "type": "div",
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "__emotion_base": "span",
      "__emotion_forwardProp": undefined,
      "__emotion_real": [Circular],
      "__emotion_styles": Array [
        Object {
          "display": "inline-block",
          "height": "24px",
          "position": "relative",
          "width": "24px",
        },
      ],
      "defaultProps": undefined,
      "displayName": "Styled(span)",
      "render": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="notification-hover-box relative"
        >
          <ModalComponent
            customClass=""
            customCloseTrigger={false}
            disabled={false}
            fullView={true}
            hasCustomClose={true}
            hasNoClose={false}
            isTriggerFullView={false}
            isWhiteIcon={false}
            onAfterModalOpen={[Function]}
            onRequestCloseCustom={[Function]}
            trigger={
              <NotificationBell
                markNotificationRead={[Function]}
                unreadNotificationCount={0}
                userId=""
              />
            }
          >
            <div
              className="notifications-head absolute wfull hfull flex flexDColumn pb48"
            >
              <ModalHeaderCustom
                heading="Notifications"
              />
              <NotificationList
                loadMoreNotifications={[Function]}
                messages={Array []}
                notificationsPresent={false}
                remainingMessages={0}
                userId=""
              />
            </div>
          </ModalComponent>
        </div>,
        "className": "flex alignCenter justifyCenter at_nofification",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ModalComponent
            customClass=""
            customCloseTrigger={false}
            disabled={false}
            fullView={true}
            hasCustomClose={true}
            hasNoClose={false}
            isTriggerFullView={false}
            isWhiteIcon={false}
            onAfterModalOpen={[Function]}
            onRequestCloseCustom={[Function]}
            trigger={
              <NotificationBell
                markNotificationRead={[Function]}
                unreadNotificationCount={0}
                userId=""
              />
            }
          >
            <div
              className="notifications-head absolute wfull hfull flex flexDColumn pb48"
            >
              <ModalHeaderCustom
                heading="Notifications"
              />
              <NotificationList
                loadMoreNotifications={[Function]}
                messages={Array []}
                notificationsPresent={false}
                remainingMessages={0}
                userId=""
              />
            </div>
          </ModalComponent>,
          "className": "notification-hover-box relative",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <div
              className="notifications-head absolute wfull hfull flex flexDColumn pb48"
            >
              <ModalHeaderCustom
                heading="Notifications"
              />
              <NotificationList
                loadMoreNotifications={[Function]}
                messages={Array []}
                notificationsPresent={false}
                remainingMessages={0}
                userId=""
              />
            </div>,
            "customClass": "",
            "customCloseTrigger": false,
            "disabled": false,
            "fullView": true,
            "hasCustomClose": true,
            "hasNoClose": false,
            "isTriggerFullView": false,
            "isWhiteIcon": false,
            "onAfterModalOpen": [Function],
            "onRequestCloseCustom": [Function],
            "trigger": <NotificationBell
              markNotificationRead={[Function]}
              unreadNotificationCount={0}
              userId=""
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <ModalHeaderCustom
                  heading="Notifications"
                />,
                <NotificationList
                  loadMoreNotifications={[Function]}
                  messages={Array []}
                  notificationsPresent={false}
                  remainingMessages={0}
                  userId=""
                />,
              ],
              "className": "notifications-head absolute wfull hfull flex flexDColumn pb48",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "heading": "Notifications",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "loadMoreNotifications": [Function],
                  "messages": Array [],
                  "notificationsPresent": false,
                  "remainingMessages": 0,
                  "userId": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
        "type": "div",
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "span",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          Object {
            "display": "inline-block",
            "height": "24px",
            "position": "relative",
            "width": "24px",
          },
        ],
        "defaultProps": undefined,
        "displayName": "Styled(span)",
        "render": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
